From 66c7ea8d98f89feff47b7c342c731900fba6aee0 Mon Sep 17 00:00:00 2001
From: Juri Lelli <juri.lelli@arm.com>
Date: Thu, 30 Apr 2015 17:35:23 +0100
Subject: [PATCH 3/3] DEBUG: arm: add cpu_capacity change tracepoint

This is useful when we want to compare cpu utilization and
cpu curr capacity side by side.

Change-Id: I5ac9660a84de44ff6de58db87565e3b704cfd7ac
Signed-off-by: Juri Lelli <juri.lelli@arm.com>
---
 arch/arm/kernel/smp.c        | 2 ++
 include/linux/sched.h        | 2 ++
 include/trace/events/power.h | 7 +++++++
 kernel/sched/fair.c          | 2 +-
 4 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/arch/arm/kernel/smp.c b/arch/arm/kernel/smp.c
index b263613..8fd6ace 100644
--- a/arch/arm/kernel/smp.c
+++ b/arch/arm/kernel/smp.c
@@ -48,6 +48,7 @@
 #include <asm/mach/arch.h>
 #include <asm/mpu.h>
 
+#include <trace/events/power.h>
 #define CREATE_TRACE_POINTS
 #include <trace/events/ipi.h>
 
@@ -730,6 +731,7 @@ static int cpufreq_callback(struct notifier_block *nb,
 			cpufreq_scale(per_cpu(l_p_j_ref, cpu),
 					per_cpu(l_p_j_ref_freq, cpu),
 					freq->new);
+		trace_cpu_capacity(capacity_curr_of(cpu), cpu);
 	}
 	return NOTIFY_OK;
 }
diff --git a/include/linux/sched.h b/include/linux/sched.h
index f7bb363..1f0b5e9 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1039,6 +1039,8 @@ struct sched_group_energy {
 	struct capacity_state *cap_states; /* ptr to capacity state array */
 };
 
+unsigned long capacity_curr_of(int cpu);
+
 struct sched_group;
 
 struct sched_domain {
diff --git a/include/trace/events/power.h b/include/trace/events/power.h
index 284244e..f4be04e 100644
--- a/include/trace/events/power.h
+++ b/include/trace/events/power.h
@@ -120,6 +120,13 @@ DEFINE_EVENT(cpu, cpu_frequency,
 	TP_ARGS(frequency, cpu_id)
 );
 
+DEFINE_EVENT(cpu, cpu_capacity,
+
+	TP_PROTO(unsigned int capacity, unsigned int cpu_id),
+
+	TP_ARGS(capacity, cpu_id)
+);
+
 TRACE_EVENT(device_pm_callback_start,
 
 	TP_PROTO(struct device *dev, const char *pm_ops, int event),
diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 934a641..9ecebc6 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -4702,7 +4702,7 @@ static long effective_load(struct task_group *tg, int cpu, long wl, long wg)
  * Returns the current capacity of cpu after applying both
  * cpu and freq scaling.
  */
-static unsigned long capacity_curr_of(int cpu)
+unsigned long capacity_curr_of(int cpu)
 {
 	return cpu_rq(cpu)->cpu_capacity_orig *
 	       arch_scale_freq_capacity(NULL, cpu)
-- 
1.9.1

